### ISSUE TITLE
Considerations for Android 6.0 support

### ISSUE THREAD

**Issue by lbdroid** [phhusson/Superuser#4]
From what I can tell, the difference between 5.x and 6.0 basically involve a tightened up sepolicy. Specifically (and most likely), things like https://android.googlesource.com/platform/external/sepolicy/+/356df32778732aa576e15071bf2736fbbd778b77%5E!/#F0

The evil closed source supersu has implemented a sepolicy patcher in their "supolicy" command. It takes the factory sepolicy file and generates a "patched" sepolicy that works with the rest of their code. HOWEVER, they refuse to even document the changes they are making to the sepolicy, which makes it impossible to trust them.

It is likely that a similar policy patch will be required. It should be possible to build it into an update zip that can repack the boot.img similar to things like http://forum.xda-developers.com/nexus-6/development/fix-fed-patcher-forceencrypt-disable-t3203867

I also suggest that consideration should be made to having the *entire* su system (save for the android apk) made part of the boot.img rather than breaking dmverity on the system partition. Since the boot.img will require modification either way, it is trivial to add the rest in there. Possibly even simplifies the process, since there are other hijacks made available for obtaining the init context from the boot.img.

**Comment by phhusson** [comment]
(This message is a Request For Comments, I want as many opinions as possible)

Doing the boot.img changes are out of context of this Superuser, though it totally kills the idea I had to line up to SuperSU to have system-only modifications.
neverallow load_policy indeed is a killer, as is dm-verity. Both will most likely be enforced by CTS/CDD.

We need greater tools to make those modifications. Now, there are many ways to go.
Those modifications can be done on the device, or offline.
This tool could just be used to root, or it can be used to do many more things.
It can meant to be just quick&dirty, or it can be meant to be maintainable and secure (at least not a blatant hole).
New Nexus-es, and all next Android phones will be required to have the /verified/ bootloader status, in which 
a user can put his own security keys, so his phone will be as secure as a locked device! This would also mean to support dm-verity resigning.

My point of view is that the rooting a device workflow should be (assuming you have an original boot.img, either a factory image or an OTA):
0. Unlock the bootloader
1. Extract the boot.img (from OTA, factory image, MTK DL tool, ...)
2. Generate a new boot.img which will backup system/recovery/boot/modem/.... (everything except user data?), with ADB enabled to have a progress bar/control (easier than to handle the screen)
3. fastboot boot (or flash if no other way...) it, wait for full backup
4. Retrieve the files
5. Generate a new boot.img which includes su daemon/appropriate SELinux rules, and it will open a web browser/play store to the APK (And will enforce the signature of the APK to prevent hijacking)
6. Check if there is a keystore partition, if so sign boot.img with appropriate key
7. fastboot boot to check
8. Generate a new recovery.img which only changes security keys, to be able to apply own update.zip
9. fastboot flash
10. Optionnally repack all of that in update.zip/factory image

Steps 1/2/3 4/5/7 can be fully automated and are really much alike, so steps 2/5 will have the same main code, only changing in the "core" part (I'd call it inBootImgPart). Steps 1/4 will most likely be actually the same (a TWRP dump looks like a factory image). Steps 3/7 are automated, but not really useful.
Considering we have such a tool, having "forceencrypt disable" would be a one-liner inBootImgPart function.

Users wanting to have their own keystore/verified/recovery keys would just need to re-launch the script.
It would be easy to automate different flavors of boot.img (root/not root, open/closed selinux, (not)encrypted, dm-verity, overclock, DTS changes, ...)

And all that, with a build bot, so new releases are fast.

Best part of this way (compared to previous ones requiring TWRP):
- No need to know how to un-encrypt /data (known to be changed by many manufacturers)
- No need to bypass dm-verity (if we want, we can even mount --bind to /system/ to replace files, or resign it by ourselves!)
- No need to have working display driver (it's usually ok on TWRP, but that's not always a given)

You only need to know how to repack a boot.img.

**Comment by lbdroid** [comment]
I do *not* think that aiming to be fully aligned with supersu is necessarily a good approach. All things should be considered in terms of what makes the tool most effective, rather than what others who have different motivations have chosen. There are some severe weaknesses in "traditional" root control infrastructure on Android, in particular, how su is treated as sudo by the authorization mechanism "application_user ALL=(ALL) ALL". Or more accurately, "sudo su". It is very weak -- the control should be much more finely grained such that each application can have commands individually authorized. Of course, this kind of changes really should be implemented later on, once there is actually something working on current Android without having to rip open severe holes in the security to do so.

Ah, you know what, looking more at your work, it seems that the policy modification tool I've been thinking about could be an extension of the sepolicy-inject tool. It would just need to support neverallow, and rule removals.

Ideally, the tools should be kept as general as possible, preferably usable both on the device as well as off. I do like the idea of tools produces being general enough to be applicable to things besides obtaining root, but first and foremost, it needs to solve the immediate problem -- root.

Usually, I'm not in favor of building quick and dirty, unless there is a commitment to revisit it and build something more resilient and useful later. I suppose that a good starting point would be a list of uses for a binary sepolicy editor.

For step 1, note that the boot.img can be obtained directly from the boot partition as well. The tool I linked to uses this approach. So effectively, an update.zip could be produced, which reads the boot.img, extracts it, modifies the sepolicy to support needs of root, adds in su and related files, rebuilds the boot.img, and writes it to the boot partition.

I'm not clear on step 5 about how a boot.img could be used to launch a web browser and obtain the superuser control application. I'm also not convinced that it is necessary. Since the process requires some manual steps, I think you could count on the user to install the application themselves.

And obviously, some of those steps will be unnecessary on devices for which there is either (a) full set of images available [nexus], or (b) no support for user-installed keys and thus no use in re-signing the sysimage. I.e., minimum 0, 1, 5, 9 (or their on-device equivalents as applicable), optional(7, 8, 10).

I could actually see the "rooting" process as this (for *current* phones);
0) unlock,
1) reboot to twrp or similar,
2) adb sideload root.zip (which reads boot partition, extract, fix sepolicy, add root files, rebuild boot.img, write to device),
3) reboot into Android,
4) install superuser.apk
... which is pretty much the same, from the user's point of view, as it was in <= 5.1.

**Comment by phhusson** [comment]
2015-10-13 23:48 GMT+02:00 lbdroid <notifications@github.com>:
>
> Ah, you know what, looking more at your work, it seems that the policy
> modification tool I've been thinking about could be an extension of the
> sepolicy-inject tool. It would just need to support neverallow, and rule
> removals.
>
I think that the best to handle such a thing, is to have every root-able
app run in their own SELinux context (sepolicy-inject can create contexts
on the fly). But it's far from easy to handle. (Link2SD even if it has
"only" rights to /data and /sdcard, should be considered to have every
rights, because it has access to system's app data)

> Ideally, the tools should be kept as general as possible, preferably
> usable both on the device as well as off. I do like the idea of tools
> produces being general enough to be applicable to things besides obtaining
> root, but first and foremost, it needs to solve the immediate problem --
> root.
>
Just the part "change a boot.img to a boot.img with su enabled" is both
trivial and complex. On a Nexus it's trivial, on a random Samsung
Spreadtrum, it might be way more complicated.
The only non-trivial part is the "opening the boot.img" part, and this
knowledge can be used to much more than just "root". I know that one can
already find on the internet tools to edit MTK boot.img, Rockchip boot.img,
etc... but I haven't found anything consistent.

Usually, I'm not in favor of building quick and dirty, unless there is a
> commitment to revisit it and build something more resilient and useful
> later. I suppose that a good starting point would be a list of uses for a
> binary sepolicy editor.
>
Yup, I'm definitely missing the full list of use cases.

> For step 1, note that the boot.img can be obtained directly from the boot
> partition as well. The tool I linked to uses this approach. So effectively,
> an update.zip could be produced, which reads the boot.img, extracts it,
> modifies the sepolicy to support needs of root, adds in su and related
> files, rebuilds the boot.img, and writes it to the boot partition.
>
This assumes you have a TWRP, which means you already have someone who did
several of hours of work, including editing boot.img. Adding su in the
boot.img would be a piece of cake for such a dev. I'm more targeting
devices with poor developers coverage, more ""official"" support to
reassure users (by specifying "this root is made for this exact firmware
and has been tested on it"), and more protected data.

Though I understand your target is more "big" devices, but many ROMs, which
doesn't have root by default.

> I'm not clear on step 5 about how a boot.img could be used to launch a web
> browser and obtain the superuser control application. I'm also not
> convinced that it is necessary. Since the process requires some manual
> steps, I think you could count on the user to install the application
> themselves.
>
Yes, I'm not convinced myself either of the real need of it.
As to how, am start -n ....VIEW -d market://... should do the trick


My aim is to have a configuration files which would look something like
that:
https://gist.github.com/phhusson/ad4ac3d910090512ab76
the openFirmware would need to be in another file, and actually be a full
list of firmwares to generate from.

For such completeness, the list of needed tools is quite high:
- dtc
- dtbtool (qcdt v1/v2/v3; sprd)
- rockchip's resource_tool
- rkcrc
- busybox obviously (for gzip, xz, cpio)
- xxxx for standard zImage-dtb (Nexus 5, MTK, ?)
- abootimg + qcdt/sprd repack
- boot_signer
- dumpkey.jar (oops, java)
- keystore_signer.jar (bis)

It is possible to handle all of them in an update.zip, but sounds annoying.
By limiting only to editing the inside of initramfs, and no key changing at
all, the number greatly decreases, but I'd like to be able to provide
additional feature without reducing security too much.

Truth is, I have a lot of devices to test an offline method, but not many
for an online one (ie not many has a working TWRP).
I'll see what I can do though.

**Comment by lbdroid** [comment]
Having every application given its own context is certainly an interesting approach. I can see the benefit of it, but it really feels like it could be a challenge to implement, and I won't even guess at what it would be to maintain as the Android selinux policies evolve. The problem I see is that the application's root context has to be something for which the target resource can communicate back to. At least that is what I think I understand of the way the selinux sandboxes are set up.

I wouldn't argue a need for, at least, certain parts to be able to work independently of the device itself, or of an off-device universal solution. I don't think that anybody could expect you, or anyone else, to build an update.zip'able solution that would work everywhere. It is probably doable to make a universal *qualcomm* update.zip, since they all use generally the same boot.img setup, and the existing tools already cater to qualcomm.

I'm going to look more into the specific needs for supporting 6.0 as soon as I get a chance (I'm on vacation at the moment).... unless you or someone else gets to it before me. Been trying to figure out what the actual policy changes are that supersu is making, but unfortunately, sediff that I have on F23 only supports policydb version 29, and Android 6.0 has a, somewhat hacked (at least, that's what I read on some nsa mailing list archive), version 30. When I get home, I'm going to try building some boot images with the more obvious selinux policy problems fixed and see where that gets me.

**Comment by lbdroid** [comment]
I'm thinking that it would be a good idea to define a new root context with all the access/control that is expected of root, and a transition to that context upon executing su/sudo. This probably would have been the correct approach to root+selinux from the beginning and would eliminate the app_process hack for holding onto the init_t context. With a correctly set up sepolicy, it should be possible to leave in the neverallow load_policy rule while providing a root that has proper control over the device.

What do you think?

**Comment by phhusson** [comment]
True. We could even drop the daemon part, and rely on good old suid (or capabilities?)

**Comment by lbdroid** [comment]
Right. That would be ideal. The daemon part seems to just add complexity once the policies are right.

Have a look at this; https://android.googlesource.com/platform/external/sepolicy/+/master/su.te

Looks like they already set up the root context, just have it disabled on user builds.

**Comment by phhusson** [comment]
Yeah well, this su is only meant to be run from adb, and is running in fully permissive mode.

**Comment by lbdroid** [comment]
Right, so we tweak it a little bit and have something useful.

**Comment by lbdroid** [comment]
Ah, I didn't notice this commit before, but it can simplify "fixing" the sepolicy;
https://android.googlesource.com/platform/external/sepolicy/+/4abd409af0e7d7fb908e5f04fa1ed946e2996dce%5E!/#F1

**Comment by lbdroid** [comment]
There are a number of additional places where changes will have to be made in order to avoid the app_process hack, such as;

No domain transitions from app:
https://android.googlesource.com/platform/external/sepolicy/+/master/app.te#271

Nobody can execute su:
https://android.googlesource.com/platform/external/sepolicy/+/master/domain.te#411

There are a bunch of places where things are conditional on userdebug_or_eng(...) that will have to be looked at.

**Comment by phhusson** [comment]
The thing is you're referring to changes in sepolicy's source, but the tool would be working on binary sepolicy, and it's a lot different.
Most major difference, is that "neverallow" rules don't go into sepolicy, it's a only a build-time verification. (and for CTS)

**Comment by lbdroid** [comment]
Maybe so, but if the repaired binary policy can be generated from source, then that is as good as (or even better) than generating it from the original binary policy. I know, not as easy as that for those devices that have significantly modified and undocumented policies, but possibly a starting point if there was some reasonable way to compare the binary policies.

**Comment by lbdroid** [comment]
Oh, I see what you're saying.... neverallow is a verification step in the compilation process, which means that it can just be added after to the binary policy without conflict.

**Comment by lbdroid** [comment]
It seems to me that sepolicy-inject is quite close to being capable of adding the rules needed. It can create a domain, and it can create allow rules. There are a few macros used in the source for the domain, from te_macros. Mostly fairly simple lists of allow rules, but a few other rules like type_transition, typeattribute, type, and dontaudit... the latter being non-critical, since all it seems to do is suppress the event from the audit log.

**Comment by lbdroid** [comment]
Ok, so some time next week when I'm home, I'm going to try fixing an sepolicy (from source) with minimum required hacks to get a working su with setuid and dead simple sh copied to su, which yes, is totally dangerous and obviously won't be staying that way for long.

**Comment by phhusson** [comment]
Ok, it's decided. I'll go (back?) to a setuid, daemon-less, /sbin/su (or /su?)/

**Comment by lbdroid** [comment]
I've reverse engineered the changes that are required for chainfire supersu (and added a few of my own along the lines needed for this recent change);

```
diff --git a/app.te b/app.te
index 40de074..98bb663 100644
--- a/app.te
+++ b/app.te
@@ -283,7 +283,7 @@ neverallow appdomain { domain -appdomain }:process
 # Transition to a non-app domain.
 # Exception for the shell domain and the su domain, can transition to runas,
 # etc.
-neverallow { appdomain -shell userdebug_or_eng(`-su') } { domain -appdomain }:process
+neverallow { appdomain -untrusted_app -shell userdebug_or_eng(`-su') } { domain -appdomain }:process
     { transition dyntransition };
 
 # Write to rootfs.
diff --git a/domain.te b/domain.te
index 0f6c6da..b1d7c41 100644
--- a/domain.te
+++ b/domain.te
@@ -396,7 +396,7 @@ neverallow domain { file_type fs_type dev_type }:{ lnk_file fifo_file sock_file
 # Nobody should be able to execute su on user builds.
 # On userdebug/eng builds, only dumpstate, shell, and
 # su itself execute su.
-neverallow { domain userdebug_or_eng(`-dumpstate -shell -su') } su_exec:file no_x_file_perms;
+neverallow { domain -init -untrusted_app userdebug_or_eng(`-dumpstate -shell -su') } su_exec:file no_x_file_perms;
 
 # Do not allow the introduction of new execmod rules. Text relocations
 # and modification of executable pages are unsafe.
diff --git a/init.te b/init.te
index 41eafe2..e7dd87a 100644
--- a/init.te
+++ b/init.te
@@ -123,7 +123,7 @@ allow init security_file:dir { create setattr };
 
 # Reload policy upon setprop selinux.reload_policy 1.
 # Note: this requires the following allow rule
-#   allow init kernel:security load_policy;
+allow init kernel:security load_policy;
 # which can be configured on a device-by-device basis if needed.
 r_dir_file(init, security_file)
 
@@ -283,4 +283,5 @@ neverallow init shell_data_file:lnk_file read;
 neverallow init app_data_file:lnk_file read;
 
 # init should never execute a program without changing to another domain.
-neverallow init { file_type fs_type }:file execute_no_trans;
+allow init { file_type fs_type }:file execute_no_trans;
+allow init kernel:security read_policy;
diff --git a/keystore.te b/keystore.te
index 83a0e85..d742d30 100644
--- a/keystore.te
+++ b/keystore.te
@@ -24,7 +24,7 @@ selinux_check_access(keystore)
 ###
 
 neverallow { domain -keystore } keystore_data_file:dir ~{ open create read getattr setattr search relabelto ioctl };
-neverallow { domain -keystore } keystore_data_file:notdevfile_class_set ~{ relabelto getattr };
+neverallow { domain -keystore -init } keystore_data_file:notdevfile_class_set ~{ relabelto getattr };
 
 neverallow { domain -keystore -init } keystore_data_file:dir *;
 neverallow { domain -keystore -init } keystore_data_file:notdevfile_class_set *;
diff --git a/su.te b/su.te
index d4a488b..1d1f6da 100644
--- a/su.te
+++ b/su.te
@@ -7,6 +7,7 @@ userdebug_or_eng(`
   # wrapped to ensure that it does not exist at all on -user builds.
   type su, domain, mlstrustedsubject;
   domain_auto_trans(shell, su_exec, su)
+  domain_auto_trans(untrusted_app, su_exec, su)
 
   # Allow dumpstate to call su on userdebug / eng builds to collect
   # additional information.
diff --git a/vold.te b/vold.te
index b22436f..fa1a879 100644
--- a/vold.te
+++ b/vold.te
@@ -164,7 +164,7 @@ allow vold self:capability sys_chroot;
 allow vold storage_file:dir mounton;
 
 neverallow { domain -vold } vold_data_file:dir ~{ open create read getattr setattr search relabelto ioctl };
-neverallow { domain -vold } vold_data_file:notdevfile_class_set ~{ relabelto getattr };
+neverallow { domain -vold -init } vold_data_file:notdevfile_class_set ~{ relabelto getattr };
 neverallow { domain -vold -init } vold_data_file:dir *;
 neverallow { domain -vold -init } vold_data_file:notdevfile_class_set *;
 neverallow { domain -vold -init } restorecon_prop:property_service set;
```

**Comment by lbdroid** [comment]
The *three* allow lines for init are ALL that is required to make supersu work;

```
+allow init kernel:security load_policy;
+allow init { file_type fs_type }:file execute_no_trans;
+allow init kernel:security read_policy;
```

I should be able to try out your latest commits momentarily.

**Comment by phhusson** [comment]
Now you no longer need to do anything besides putting su in boot.img and chmod +s

**Comment by lbdroid** [comment]
Would also need a chcon su_exec, no?

**Comment by phhusson** [comment]
Well, it would need a lot of changes in sepolicy, not just those three lines. (or you're using eng su.te as well?)
It feels like Chainfire's SuperSU is still using daemon (hence it's init who launches it).
You'd rather want to allow all domains to execute su_exec

FIY, ATM i'm developping on a !enforcing device. I'll start making full list of sepolicy-inject soon.

**Comment by lbdroid** [comment]
I'm building my boot.img as userdebug, so yeah, its coming with su/su_exec. I am working in enforcing right now though.

**Comment by lbdroid** [comment]
Ah, /file_contexts
/sbin(/.*)?		u:object_r:rootfs:s0

What Android version are you working with now?

**Comment by phhusson** [comment]
CAF 6.0

**Comment by lbdroid** [comment]
system/core/libcutils/fs_config.c is where to set the file permissions to 06755.

**Comment by phhusson** [comment]
Erf, apps not being able to run a set-uid app is not an SELinux limitation...
I'll try to find what it is and hope it's not too hard to circumvent.

**Comment by lbdroid** [comment]
I was noticing that too, but read something about zygote spawned processes being nosuid on the /system partition, so I figured you ran into that because you've already jumped to the step of su being moved to the boot partition. Hopefully it isn't something more universal that gets slapped on at the userspace end of things.

**Comment by phhusson** [comment]
It's PR_SET_NO_NEW_PRIVS (cf http://man7.org/linux/man-pages/man2/prctl.2.html)
I still haven't found who sets it in Android framework, but that doesn't sound good...

**Comment by lbdroid** [comment]
Well, I suppose it could be sliced off the kernel, but that is probably about 10 times more radical of a hack than we should be aiming for.

**Comment by lbdroid** [comment]
There's a heap of references to it in the CTS tests.

**Comment by phhusson** [comment]
First line of app_process.c. Should be bypassable with a LD_PRELOAD. So it's a choice between ugly LD_PRELOAD and reverting to daemon...

**Comment by lbdroid** [comment]
Hmm. It would probably be most flexible to support both options. Maybe a compile-time flag that builds su without the daemon parts?

**Comment by phhusson** [comment]
Well, I don't like having multiple untested codepaths, but it's just three lines in su_main, so I can keep it as is.

**Comment by lbdroid** [comment]
My initial reaction when you mentioned LD_PRELOAD was that it would be way to ugly and hackish. But I've been thinking about it for a bit and it seems to me that it would actually create a real good opportunity to maintain a hardened security while making a more conventional root possible.

PR_SET_NO_NEW_PRIVS controls the behavior of execve. So we *also* intercept execve, and can, at *that* point, decide whether to allow setuid on the su binary, or even choose to apply PR_SET_NO_NEW_PRIVS against the child under certain circumstances.

**Comment by lbdroid** [comment]
It would actually be incredibly easy to set it up.
Add a line to init.zygote.rc:
setenv LD_PRELOAD /sbin/libsu.so

**Comment by lbdroid** [comment]
I have it working with selinux permissive using daemon mode for now, and all from the boot.img. Definitely a good step.

**Comment by lbdroid** [comment]
My mission tomorrow is going to be to get the sepolicy together so that I can set it back to enforcing. Then I'll see what I can do about an LD_PRELOAD hack.

**Comment by lbdroid** [comment]
I'm keeping a patch for building this into a boot image from source here;
https://github.com/lbdroid/AOSP-SU-PATCH

I will keep it up to date with progress.

**Comment by lbdroid** [comment]
The program, yes. But the program can't do that until after its been exec'd. By the time the program is in a state where it can mess with LD_PRELOAD, we've already shoved PR_SET_NO_NEW_PRIVS down its throat.

It would make the authentication process a little bit more cumbersome, since we would proactively apply that limitation before the process has the opportunity to request SU permissions, which means that it wouldn't apply until the process is killed and relaunched.

If the program has already been authenticated for root access, then this at least doesn't add *more* of a security hole than it already has with root.

The other thing to consider, is that PR_SET_NO_NEW_PRIVS doesn't really open up that much of a security hole to begin with, especially on a device running dm-verity on the system partition. There just aren't any setuid/gid binaries to take advantage of.

**Comment by phhusson** [comment]
I don't understand what you want to do with LD_PRELOAD.
Anyway, I totally agree that it's not a big threat with dm-verity:
Putting su in boot.img makes dm-verity still intact
And we can keep SELinux rule which prevents overwriting boot.img
My problem is more the likelyhood of breakage (redirecting calls to libc by a simple LD_PRELOAD works only on >= 6.0 AFAIK, I'd like to still work on older linkers)

That makes me think we might want to have a neverallow-like feature, where we detect when an app requests write access to boot.img, and show a big-ass warning.

**Comment by lbdroid** [comment]
Nevermind this line of thought. I don't think it can be implemented cleanly, and not without hacks within the system partition. app_process doesn't fork. It gets called separately for every application being launched, which means that we don't have full control over the environment prior to it running.

**Comment by lbdroid** [comment]
I think I'm going to let the non-daemon possibilities sit on the back burner for a while. There is no obvious way to make that work right without rebuilding the kernel itself... not that that is such a bad thing for later on, but it is a distraction right now in terms of getting things working properly with selinux enforcing.

**Comment by phhusson** [comment]
AFAIK, LD_PRELOAD=/libsu.so should work to prevent app_process to call PR_SET_NO_NEW_PRIVS, without changing kernel nor /system

**Comment by lbdroid** [comment]
Right, but how to you enforce that environment variable globally to processes that are forked from something sitting in the system partition?

**Comment by phhusson** [comment]
I can't see why I would need that? I only need it for app_process which is run by init.rc

**Comment by lbdroid** [comment]
That's my point. init.rc only runs the system-server *instance* of app_process. Every single application runs app_process again.

Here is init.zygote32.rc:
```
service zygote /system/bin/app_process -Xzygote /system/bin --zygote --start-system-server
    class main
    socket zygote stream 660 root system
    onrestart write /sys/android_power/request_state wake
    onrestart write /sys/power/state on
    onrestart restart media
    onrestart restart netd
```

Look at frameworks/base/cmds/app_process/app_main.cpp from 246, notice it has additional parameters, like --application? This file is not just run once. It is run once for *every* application.

**Comment by lbdroid** [comment]
Here is the main problem;
if you run the hack on the system-server instance of app_process, you don't gain anything, since system-server doesn't need to be granted root access.

The instances that need setuid are the --application instances, which correspond to actual software that the user is going to be running and needs to setuid, like a terminal emulator.

**Comment by lbdroid** [comment]
I need to add a patch to system/core/init/init.c to restorecon on the su binary.

Another way to do it is within an init script using something like this;
mount -o remount,rw /
restorecon /sbin/su
mount -o remount,ro /

Otherwise, this binary will *always* have the context set to u:object_r:rootfs:s0, which is obviously not very helpful for automatic transitions.

**Comment by phhusson** [comment]
Oops, you're right, thanks.
Now, it's not the app who calls app_process, I guess it's system_server who does it, so it would still work (unless it explicitly forgets LD_PRELOAD)

**Comment by phhusson** [comment]
Changing file_contexts should be enough to change context of /sbin/su, no?

**Comment by lbdroid** [comment]
No, the ramdisk doesn't store the extended attributes, which means that it has to be restored after selinux is initialized. I spent about 3 hours bashing my head against a wall last night figuring this out.

**Comment by phhusson** [comment]
Yes, but file_contexts is the file read at run-time to apply context

**Comment by lbdroid** [comment]
Right, but to make it apply, you still need to run restorecon, which it does *NOT* do automatically.

**Comment by lbdroid** [comment]
http://blog.gmane.org/gmane.comp.security.seandroid/month=20150601
http://comments.gmane.org/gmane.comp.security.seandroid/2301

The first link is a discussion about adding built-in file context support to the rootfs, the second is discussing adding a restorecon_recursive("/sbin") to init.c.

**Comment by lbdroid** [comment]
Actually may not have to go to that extreme, since with daemon mode the application doesn't need to make the transition, just the daemon, and that can be transitioned using seclabel from the init.zygote

**Comment by phhusson** [comment]
Thanks for all your research, nice to know.
Having a transition when calling /sbin/su would be great to have better protection on the su socket/pts, and I don't see a problem in doing remount rw/restorecon/remount ro. I don't think manufacturer would break that.

But I agree that for first steps seclabel is enough.

**Comment by lbdroid** [comment]
Good point!

**Comment by lbdroid** [comment]
I've managed to get su working through adb (shell context), having selinux enforcing.

**Comment by phhusson** [comment]
Nice :-)
Next step is from untrusted_app

**Comment by lbdroid** [comment]
Correction: I actually have it working also from untrusted_app, there are just some context screwups that limit what I can do with it.

If I try using the --context option, it gives a hard time connecting to the daemon.

**Comment by lbdroid** [comment]
Like this:
```
[  508.008088] type=1400 audit(1445529993.672:76): avc: denied { getattr } for pid=5394 comm="sh" path="/dev/pts/5" dev="devpts" ino=8 scontext=u:r:system_app:s0 tcontext=u:object_r:untrusted_app_devpts:s0:c512,c768 tclass=chr_file permissive=0
[  508.008677] type=1400 audit(1445529993.672:77): avc: denied { ioctl } for pid=5394 comm="sh" path="/dev/pts/5" dev="devpts" ino=8 ioctlcmd=5401 scontext=u:r:system_app:s0 tcontext=u:object_r:untrusted_app_devpts:s0:c512,c768 tclass=chr_file permissive=0
[  508.008922] type=1400 audit(1445529993.672:78): avc: denied { getattr } for pid=5394 comm="sh" path="/dev/pts/5" dev="devpts" ino=8 scontext=u:r:system_app:s0 tcontext=u:object_r:untrusted_app_devpts:s0:c512,c768 tclass=chr_file permissive=0
[  508.009156] type=1400 audit(1445529993.672:79): avc: denied { read } for pid=5394 comm="sh" path="/dev/pts/5" dev="devpts" ino=8 scontext=u:r:system_app:s0 tcontext=u:object_r:untrusted_app_devpts:s0:c512,c768 tclass=chr_file permissive=0
```

**Comment by phhusson** [comment]
I guess it's acceptable to whitelist all contexts access to untrusted_app_devpts, since we still have unix permissions... ? (only problem would be with dac_override)

**Comment by phhusson** [comment]
Or su daemon could whitelist on-the-fly, when su --context XXXX is called.

**Comment by lbdroid** [comment]
This is an interesting one, running --context u:r:su:s0 yields root, then running "dumpsys power" it spits out this;

```
[  690.664395] type=1400 audit(1445530176.322:80): avc: denied { read write } for pid=5679 comm="dumpsys" path="/dev/pts/5" dev="devpts" ino=8 scontext=u:r:system_server:s0 tcontext=u:object_r:untrusted_app_devpts:s0:c512,c768 tclass=chr_file permissive=0
```

Note that that is from system_server, which I wasn't expecting to see.

**Comment by lbdroid** [comment]
Can it be whitelisted on the fly without reloading the policy? If possible, it would be nice to be able to block policy reloads like google has it, since one bad application could, in theory, completely destroy the security of the device if it is able to obtain that kind of access.

**Comment by lbdroid** [comment]
Hmm, looks like the auto transition isn't happening from untrusted_app.

```
u:r:untrusted_app:s0:c512,c768 u0_a124   6773  6083  s
```

May have something to do with the categories?

**Comment by phhusson** [comment]
No, you obviously need to reload policy to whitelist, that's the problem.
My guess is google no longer has the ability to reload policy, since they removed load_policy from every context.

What do you get in audit log?

**Comment by lbdroid** [comment]
Nothing at all in the audit log regarding the transition.

**Comment by phhusson** [comment]
Do you have any dontaudit in your policy?

**Comment by lbdroid** [comment]
Default policy is loaded with dontaudit's. I'm looking at them right now.

**Comment by lbdroid** [comment]
May not actually be correct that the auto transition isn't working. Looks like some of these things aren't actually reported from ps -Z.

For instance, from adb, I ran "su --context u:r:system_app:s0", yet from a *different* adb, it still shows as u:r:su:s0, but when I run things from the one on system_app, it appears to be restricted to system_app anway.

```
# ps -Z                                                           
LABEL                          USER      PID   PPID  NAME
u:r:system_app:s0              system    1651  320   com.quicinc.cne.CNEService
u:r:system_app:s0              system    2716  320   com.qualcomm.telephony
u:r:system_app:s0              root      6997  6992  sh
u:r:system_app:s0              root      7110  6997  ps
```

**Comment by phhusson** [comment]
Can you check you can read /proc/<pid>/attr/current and that it contains the same result as ps -Z?

**Comment by lbdroid** [comment]
No such file or directory. Remember that I'm running factory nexus 6 kernel.

**Comment by lbdroid** [comment]
Ah, interesting.

su --context u:r:system_server:s0              
Cannot execute /system/bin/sh: Permission denied

Which is good. It means that I have, indeed, entered the system_server domain.

**Comment by phhusson** [comment]
sorry I meant /proc/ PID /attr/current

**Comment by lbdroid** [comment]
It matches ps -Z

**Comment by lbdroid** [comment]
```
$ grep -rn untrusted_app_devpts *
grep: Android.bp: No such file or directory
grep: bootstrap.bash: No such file or directory
Binary file out/target/product/shamu/root/sepolicy matches
Binary file out/target/product/shamu/obj/ETC/sepolicy_intermediates/sepolicy matches
out/target/product/shamu/obj/ETC/sepolicy_intermediates/policy.conf:3843:type untrusted_app_devpts, fs_type;
out/target/product/shamu/obj/ETC/sepolicy_intermediates/policy.conf:3847:type_transition untrusted_app devpts:chr_file untrusted_app_devpts;
out/target/product/shamu/obj/ETC/sepolicy_intermediates/policy.conf:3851:allow untrusted_app untrusted_app_devpts:chr_file { open getattr read write ioctl };
out/target/product/shamu/obj/ETC/sepolicy_intermediates/policy.conf.dontaudit:3843:type untrusted_app_devpts, fs_type;
out/target/product/shamu/obj/ETC/sepolicy_intermediates/policy.conf.dontaudit:3847:type_transition untrusted_app devpts:chr_file untrusted_app_devpts;
out/target/product/shamu/obj/ETC/sepolicy_intermediates/policy.conf.dontaudit:3851:allow untrusted_app untrusted_app_devpts:chr_file { open getattr read write ioctl };
Binary file out/target/product/shamu/obj/ETC/sepolicy_intermediates/sepolicy.dontaudit matches
```
I'm inclined to open that up to everything. Its already opened up to the least trustworthy domain in existence, untrusted_app...

**Comment by phhusson** [comment]
Which actually makes sense, since it's the app who creates it -_-'

**Comment by lbdroid** [comment]
Right.

**Comment by lbdroid** [comment]
allow domain untrusted_app_devpts:chr_file { open getattr read write ioctl };

Probably can narrow that down at some point.

**Comment by lbdroid** [comment]
Hmm, well closer now...
If I don't set the context, or leave the context alone, I can now successfully run "dumpsys power" from terminal emulator. If I set the context to system_app, it dies with this;

```
<36>[  126.161802] type=1400 audit(1445535147.549:72): avc: denied { getattr } for pid=4741 comm="sh" path="/dev/pts/3" dev="devpts" ino=6 scontext=u:r:system_app:s0 tcontext=u:object_r:untrusted_app_devpts:s0:c512,c768 tclass=chr_file permissive=0
<36>[  126.162859] type=1400 audit(1445535147.549:73): avc: denied { getattr } for pid=4741 comm="sh" path="/dev/pts/3" dev="devpts" ino=6 scontext=u:r:system_app:s0 tcontext=u:object_r:untrusted_app_devpts:s0:c512,c768 tclass=chr_file permissive=0
<36>[  126.178793] type=1400 audit(1445535147.569:74): avc: denied { dac_override } for pid=4741 comm="sh" capability=1 scontext=u:r:system_app:s0 tcontext=u:r:system_app:s0 tclass=capability permissive=0
<36>[  126.194515] type=1400 audit(1445535147.579:75): avc: denied { write } for pid=4741 comm="sh" path="/dev/pts/3" dev="devpts" ino=6 scontext=u:r:system_app:s0 tcontext=u:object_r:untrusted_app_devpts:s0:c512,c768 tclass=chr_file permissive=0
<36>[  126.194667] type=1400 audit(1445535147.579:76): avc: denied { read } for pid=4741 comm="sh" path="/dev/pts/3" dev="devpts" ino=6 scontext=u:r:system_app:s0 tcontext=u:object_r:untrusted_app_devpts:s0:c512,c768 tclass=chr_file permissive=0
```

I'm thinking that system_app may not be a member of domain...

**Comment by lbdroid** [comment]
Its the category.

I don't know why it works with system_server and not system_app, but when I use --context "u:r:system_app:s0:c512,c768" it works beautifully -- except for dac_override, for which I don't see any pressing need to permit.

I'm pretty sure that the auto transition is failing on the same category issue.

**Comment by phhusson** [comment]
Ok thanks. I know pretty much nothing about categories, I'll try to understand what those are. (I guess that here it's used to separate apps from others?)

**Comment by lbdroid** [comment]
From what I can tell, they seem to, at least on Android, be set universally to c512,c768. As I understand it, this actually implies c0,c512,c768

**Comment by phhusson** [comment]
Ok, from what I'm reading ( https://wiki.gentoo.org/wiki/SELinux/Tutorials/SELinux_Multi-Level_Security ), categories are totally orthogonal to other parts of SELinux, and if two systems doesn't share a common category, they can't communicate. So I guess just setting c512 would be enough? (or 768)

**Comment by lbdroid** [comment]
Should be. Or a range c0.cwhatever.

**Comment by lbdroid** [comment]
Heh, sit here bashing my head against a wall for an hour trying to figure out why something isn't working and there's no audit message... two lines up an execv. Duh.

**Comment by lbdroid** [comment]
Yeah, there we are. Now I have it *starting* in enforcing, rather than switching there after booting.

**Comment by lbdroid** [comment]
FYI: I tried out the categories a bit, and turns out they work on an AND basis.

So if the resource you want to access has c512,c768, you need to match them *both*.
Looks like it has to match *precisely* also, so throwing in a c123 along with the other two doesn't work.

**Comment by lbdroid** [comment]
I think that the category problem might be with *just* terminal emulator. Reason it seems that way is that I have another program I wrote that disables and kills some services on bootup, which has to use the same system_app context (well maybe it doesn't have to any more...) but doesn't have the problem with the pts. It just works without complaint.

**Comment by lbdroid** [comment]
I've updated my patch here: https://github.com/lbdroid/AOSP-SU-PATCH

There is one additional change that will be necessary for running restorecon from a script, if you notice line 104 of the patch, it is necessary to authorize init for relabelto su_exec.

Current state is: enforcing, with policy reload disabled.

The policy patches, **FOR USERDEBUG** are pretty straightforward. I'll have a look at user build policies tomorrow.

**Comment by lbdroid** [comment]
This is the actual request window;

[ 4208.208968] type=1400 audit(1445560975.200:36): avc: denied { write } for pid=15987 comm="Thread-745" name=".socket15976" dev="tmpfs" ino=107615 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:device:s0 tclass=sock_file permissive=0
[ 4215.717367] type=1400 audit(1445560982.710:37): avc: denied { getattr } for pid=14729 comm="ruser:superuser" path="/dev/com.koushikdutta.superuser/.socket15976" dev="tmpfs" ino=107615 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:device:s0 tclass=sock_file permissive=0

**Comment by lbdroid** [comment]
This appears to be an instance of creating a socket, which inherits its parent's context, and then being locked out as a result.

# Sockets under /dev/socket that are not specifically typed.
neverallow appdomain socket_device:sock_file write;

This might be a good application for a filetrans_pattern rule.

**Comment by lbdroid** [comment]
Its another problem with these stupid categories.
u:r:untrusted_app:s0:c512,c768
u:r:foo:s0
Even when there is an allow rule for untrusted_app --> foo, foo is category c0
u:r:foo:s0 == u:r:foo:s0:c0
But because u:r:untrusted_app:s0:c512,c768 explicitely does NOT include c0, it therefore can't access the resource with label foo.

**Comment by lbdroid** [comment]
There might be a better way to do this; have an android/java service listening for connections from the su client binary, rather than the su binary waiting on the application to call it back. It can still be launched with an am start.

**Comment by lbdroid** [comment]
Or an easier way...

mlstrustedobject = ignore category when dealing with this object.
mlstrustedsubject = ignore this process's category always.

I added to the sepolicy as follows;

type su_device, dev_type, mlstrustedobject;
allow untrusted_app su_device:sock_file rw_file_perms;
allow untrusted_app su_device:dir rw_file_perms;
allow untrusted_app su_device:dir r_dir_perms;
(I think the dir rw_file_perms is probably unnecessary)

And then chcon u:object_r:su_device:s0 on /dev/{the su client's path}

Then the auth window started working.

**Comment by lbdroid** [comment]
filetrans_pattern(unconfined_t, admin_home_t, dir, ssh_home_t, “.ssh”)

first param: calling process context,
second param: parent directory context (/dev is u:object_r:device:s0),
third param: type of object being created (dir is correct),
fourth param: new context to apply to the directory (u:object_r:su_device:s0),
fifth param: the *name* of the what is being created, which seems to come from the package name.

**Comment by lbdroid** [comment]
I'm having no luck with the auto transitions. I'm thinking that I may have the calling process context wrong.

Ah, I get it... its not untrusted_app, its actually making the transition as su...

**Comment by lbdroid** [comment]
We are in *really* good shape where it comes to the source-based policy modifications, and other adjustments required for building the boot.img from source.

I've hardened up the policies by disabling policy reload, and setting enforcement for the su domain. Google was nice enough to provide a heap of dontaudit rules in the su.te that were easily converted into allow rules (except for messing with kernel security, that isn't allowed and will be blocked). Obviously, su had to be removed from a whole bunch of neverallow's to let the allows go through, but that is appropriate given what su is for -- it needs to have access to almost everything.

Naturally, the neverallow adjustments can be ignored for binary policy patching. The number of extra rules required is actually quite short.

**Comment by lbdroid** [comment]
These are the additions to the policy (outside of neverallow's) for userdebug builds.  A lot of these changes are for debugging purposes and not actually related to root.

adbd.te: (first block appears to be for "adb root", second for debugging, neither should be required)
```
userdebug_or_eng(`
  allow adbd self:process setcurrent;
  allow adbd su:process dyntransition;
')
userdebug_or_eng(`
  # Write debugging information to /data/adb
  # when persist.adb.trace_mask is set
  # https://code.google.com/p/android/issues/detail?id=72895
  allow adbd adb_data_file:dir rw_dir_perms;
  allow adbd adb_data_file:file create_file_perms;
')
```

app.te: (these appear to be debugging related)
```
# Notify zygote of the wrapped process PID when using --invoke-with.
userdebug_or_eng(`
  allow appdomain zygote:fifo_file write;
')
# Send heap dumps to system_server via an already open file descriptor
# % adb shell am set-watch-heap com.android.systemui 1048576
# % adb shell dumpsys procstats --start-testing
# debuggable builds only.
userdebug_or_eng(`
  allow appdomain heapdump_data_file:file append;
')
```

debuggerd.te: (debugging)
```
userdebug_or_eng(`
  allow debuggerd input_device:dir r_dir_perms;
  allow debuggerd input_device:chr_file rw_file_perms;
')
```

domain.te:
```
userdebug_or_eng(`
  # Same as adbd rules above, except allow su to do the same thing
  allow domain su:unix_stream_socket connectto;
  allow domain su:fd use;
  allow domain su:unix_stream_socket { getattr getopt read write shutdown };

  binder_call({ domain -init }, su)

  # Running something like "pm dump com.android.bluetooth" requires
  # fifo writes
  allow domain su:fifo_file { write getattr };

  # allow "gdbserver --attach" to work for su.
  allow domain su:process sigchld;

  # Allow writing coredumps to /cores/*
  allow domain coredump_file:file create_file_perms;
  allow domain coredump_file:dir ra_dir_perms;
')
```

init.te: (logging / debug)
```
userdebug_or_eng(`
  domain_auto_trans(init, logcat_exec, logd)
')
```

logd.te: (logging/debug)
```
# case where logpersistd is actually logcat -f in logd context (nee: logcatd)
userdebug_or_eng(`
  unix_socket_connect(logd, logdr, logd)
')
```

perfprofd.te: (performance profiling -- debug)
```
userdebug_or_eng(`
  type perfprofd, domain, mlstrustedsubject;
  init_daemon_domain(perfprofd)
  # perfprofd needs to control CPU hot-plug in order to avoid kernel
  # perfevents problems in cases where CPU goes on/off during measurement;
  # this means read access to /sys/devices/system/cpu/possible
  # and read/write access to /sys/devices/system/cpu/cpu*/online
  allow perfprofd sysfs_devices_system_cpu:file rw_file_perms;
  # perfprofd checks for the existence of and then invokes simpleperf;
  # simpleperf retains perfprofd domain after exec
  allow perfprofd system_file:file rx_file_perms;
  # perfprofd reads a config file from /data/data/com.google.android.gms/files
  allow perfprofd app_data_file:file r_file_perms;
  allow perfprofd app_data_file:dir search;
  allow perfprofd self:capability { dac_override };
  # perfprofd opens a file for writing in /data/misc/perfprofd
  allow perfprofd perfprofd_data_file:file create_file_perms;
  allow perfprofd perfprofd_data_file:dir rw_dir_perms;
  # perfprofd uses the system log
  read_logd(perfprofd);
  write_logd(perfprofd);
  # perfprofd inspects /sys/power/wake_unlock
  wakelock_use(perfprofd);
  # simpleperf uses ioctl() to turn on kernel perf events measurements
  allow perfprofd self:capability sys_admin;
  # simpleperf needs to examine /proc to collect task/thread info
  r_dir_file(perfprofd, domain)
  # simpleperf needs to access /proc/<pid>/exec
  allow perfprofd self:capability { sys_resource sys_ptrace };
  # simpleperf needs open/read any file that turns up in a profile
  # to see whether it has a build ID
  allow perfprofd exec_type:file r_file_perms;
  # simpleperf is going to execute "sleep"
  allow perfprofd toolbox_exec:file x_file_perms;
  # needed for simpleperf on some kernels
  allow perfprofd self:capability ipc_lock;
')
```

procrank.te: (I don't even know what procrank does, but it isn't required for su)
```
userdebug_or_eng(`
  type procrank, domain, mlstrustedsubject;

  domain_auto_trans(shell, procrank_exec, procrank)
  domain_auto_trans(dumpstate, procrank_exec, procrank)
  allow procrank self:capability sys_ptrace;
  allow procrank devpts:chr_file { read write getattr ioctl };
  allow procrank dumpstate:unix_stream_socket { read write getattr };
  r_dir_file(procrank, domain)
  allow procrank { shell dumpstate }:fd use;
  allow procrank adbd:process sigchld;
  # allow procrank write to bugreport.
  allow procrank shell_data_file:file w_file_perms;
')
```

su.te: (the meat)
```
userdebug_or_eng(`
  # Domain used for su processes, as well as for adbd and adb shell
  # after performing an adb root command.  The domain definition is
  # wrapped to ensure that it does not exist at all on -user builds.
  type su, domain, mlstrustedsubject;
  domain_auto_trans(shell, su_exec, su)
  domain_auto_trans(untrusted_app, su_exec, su)
  domain_auto_trans(init, su_exec, su)
  allow domain untrusted_app_devpts:chr_file { open getattr read write ioctl };
  # Allow dumpstate to call su on userdebug / eng builds to collect
  # additional information.
  domain_auto_trans(dumpstate, su_exec, su)
  # Make sure that dumpstate runs the same from the "su" domain as
  # from the "init" domain.
  domain_auto_trans(su, dumpstate_exec, dumpstate)
  # Add su to various domains
  net_domain(su)
  app_domain(su)
  allow su self:capability_class_set *;
  #allow su kernel:security *;
  allow su kernel:system *;
  allow su self:memprotect *;
  allow su { domain -kernel }:process *;
  allow su domain:fd *;
  allow su domain:dir *;
  allow su domain:lnk_file *;
  allow su domain:{ fifo_file file } *;
  allow su domain:socket_class_set *;
  allow su domain:ipc_class_set *;
  allow su domain:key *;
  allow su fs_type:filesystem *;
  allow su {fs_type dev_type file_type}:dir_file_class_set *;
  allow su node_type:node *;
  allow su node_type:{ tcp_socket udp_socket rawip_socket } *;
  allow su netif_type:netif *;
  allow su port_type:socket_class_set *;
  allow su port_type:{ tcp_socket dccp_socket } *;
  allow su domain:peer *;
  allow su domain:binder *;
  allow su property_type:property_service *;
  allow su service_manager_type:service_manager *;
  allow su keystore:keystore_key *;
  allow su domain:debuggerd *;
  allow su domain:drmservice *;
  allow su unlabeled:filesystem *;
')
```

uncrypt.te: (debug)
```
userdebug_or_eng(`
  # For debugging, allow /data/local/tmp access
  r_dir_file(uncrypt, shell_data_file)
')
```

untrusted_app.te: (debug)
```
# Allow GMS core to access perfprofd output, which is stored
# in /data/misc/perfprofd/. GMS core will need to list all
# data stored in that directory to process them one by one.
userdebug_or_eng(`
  allow untrusted_app perfprofd_data_file:file r_file_perms;
  allow untrusted_app perfprofd_data_file:dir r_dir_perms;
')
```

wpa.te: (root + wireless = required)
```
# Allow wpa_cli to work. wpa_cli creates a socket in
# /data/misc/wifi/sockets which wpa supplicant communicates with.
userdebug_or_eng(`
  unix_socket_send(wpa, wpa, su)
')
```

**Comment by lbdroid** [comment]
Add to that, of course, and the rest of the changes in my patch.

**Comment by lbdroid** [comment]
I think it may be useful to create multiple degrees of su-like domains, or at least one more in particular, with a strong commandline-only authentication mechanism that requires you to already *be* in the standard su domain.

I.e., an su-permissive domain, which can be transitioned into via something like an su-permissive_exec, and that executable with su-permissive_exec (for which su domain must be denied relabelto) will enforce some kind of strong authentication.

This would provide a semi-trusted root suitable for most uses and automatic responses, but also an extra strong one for special circumstances that would allow policy reloading, rewriting sensitive partitions like boot and recovery, and other extremely dangerous functions.

**Comment by phhusson** [comment]
So i've created https://github.com/phhusson/super-bootimg
With "install-su.sh" script, and several ndk-buildable binaries
Currently the policy I put is put domains in permissive mode.

I did it as a quick run and tested on only one device (CAF 6.0), but su is here, daemon started by init and sepolicy is changed.
I expect the sepolicy-inject commands not to be enough to work on a user build, I'll check that on hammerhead tomorrow.

**Comment by phhusson** [comment]
Works on official hammerhead M factory image, with enforcing.
Many domains are still in permissive (especially untrusted_app...), Next target is to include your changes.

**Comment by lbdroid** [comment]
The allows I added to su.te should take care of the majority of those logs.

**Comment by phhusson** [comment]
I'm hitting the c512 thingy again. I'm not seeing any reference to "type su_device, dev_type, mlstrustedobject;" in your full dump, is that still relevant?

**Comment by lbdroid** [comment]
Hmm?
Its right here; https://github.com/lbdroid/AOSP-SU-PATCH/blob/master/su.patch#L328

Don't confuse what I put into this thread with the patch. What I put into this thread was specifically the difference that google implemented between user and userdebug. At least aside from the su.te part, since I pulled those differences out of my working tree.

**Comment by phhusson** [comment]
+allow untrusted_app su_exec:file execute_no_trans;
Could you change that so it transitions to another context (su_client), instead of staying in the same context?
(That prevents su socket fuzzing)

**Comment by lbdroid** [comment]
Getting a transition working for that is on my todo list. I have it filed in an issue on my side.

**Comment by phhusson** [comment]
Ok. I keep my current su_client model, and I work on su from adb shell. I'll get back to untrusted_app when you have it.

**Comment by lbdroid** [comment]
My focus for the transition was to switch untrusted_app --> su, which would be for daemon free operation (assuming we can come up with a way to setuid). Because that isn't going to be an easy thing, I wasn't really focused on it right now. Is there a security issue with using the execute_no_trans for making the socket work?

**Comment by phhusson** [comment]
As for everything with SELinux, it's never a direct security issue. Here it is better to have a su_client context because this way, we won't have some apps trying to send weird stuff on the socket. I agree it's not a major problem, but if it's possible, I'd rather have it.

**Comment by lbdroid** [comment]
Ah, ok, I understand, keep criminals from bypassing the door by going through the window.

**Comment by lbdroid** [comment]
Wait a second though, its the application that goes into the socket, not the client binary... If we put the socket into a context that only the client binary can access, then the application will be locked out.

**Comment by phhusson** [comment]
I was only thinking about the su client <=> su daemon part. I must admit I hadn't thought of su daemon <=>  SU app yet.

**Comment by lbdroid** [comment]
Oh wait, there's two sockets... That would be fine then. Lock everything, including the application, out of the daemon socket, leave the authorization socket (which is still run by the daemon) open for authorizations. Is there any security on the authorization socket?

**Comment by lbdroid** [comment]
I'm at a loss on this one.
I have;
domain_auto_trans(untrusted_app, su_exec, su)

su is like this;
-rwsr-sr-x root     root              u:object_r:su_exec:s0 su

And I have an audit log that says this;
[   53.255419] type=1400 audit(1445797827.942:10): avc: denied { execute_no_trans } for pid=4387 comm="sh" path="/sbin/su" dev="rootfs" ino=7807 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:su_exec:s0 tclass=file permissive=0

su_exec is even defined like this;
type su_exec, exec_type, file_type, mlstrustedobject;

**Comment by phhusson** [comment]
You dropped every noaudit and haven't seen anything related?

**Comment by lbdroid** [comment]
There are a few noaudits, but not related. In any case, if it was an active denial, allowing execute_no_trans shouldn't fix it.

**Comment by lbdroid** [comment]
Its like the rule isn't actually being created.

**Comment by phhusson** [comment]
if ((bprm->file->f_path.mnt->mnt_flags & MNT_NOSUID) ||                                                                 
        (bprm->unsafe & LSM_UNSAFE_NO_NEW_PRIVS))                                                                           
        new_tsec->sid = old_tsec->sid;

**Comment by lbdroid** [comment]
What is that?

**Comment by phhusson** [comment]
Code in SELinux, if you did NO_NEW_PRIVS, you can't transition.

**Comment by lbdroid** [comment]
Bloody hell.

**Comment by phhusson** [comment]
I'm really happy, because I did all the changes needed to have unpermissive su_client... time to revert...

**Comment by lbdroid** [comment]
Right, that isn't going to be of much help, since if we can bypass that problem, we can go daemon free and transition straight to su. Hope you didn't spend too much time on it.

**Comment by phhusson** [comment]
Transitioning to SU wouldn't make us root.

**Comment by lbdroid** [comment]
Right, but setuid would.

**Comment by lbdroid** [comment]
Since it is apparently the same problem blocking setuid and selinux transitions.

**Comment by lbdroid** [comment]
If we can secure the data on the socket, that should be enough to prevent baddies from taking advantage of it. mlstrustedobject shouldn't be a big deal as long as the only thing we are adding the trust to is the socket, which it is, and which will be secure.

Maybe adding crypto to the socket. Put a crypto key beside the su binary that is unreadable to everything besides root and the su context?

**Comment by phhusson** [comment]
No need to add any crypto. Though I'll add a check of the apk's key.
And yes, the data on the socket is supposed to be secure, so I have not much problem with this method.
I'm actually currently implementing it

**Comment by lbdroid** [comment]
Ok, sounds reasonable.

**Comment by phhusson** [comment]
Ok I'm lost, ATM I get:
10-25 21:21:55.683  3947  3947 W su      : type=1400 audit(0.0:40): avc: denied { connectto } for path=005355504552555345520000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:r:su_daemon:s0 tclass=unix_stream_socket permissive=0

How do I bypass that?

**Comment by lbdroid** [comment]
That depends... what is 'su_daemon'?

**Comment by lbdroid** [comment]
Is that equivalent to my su_device?

**Comment by phhusson** [comment]
No, it is equivalent to su. I have another su_device, but this socket doesn't refer to /dev/, so I don't think it's linked to su_device?

**Comment by phhusson** [comment]
Perhaps you moved back to file-based socket instead of abstract socket?

**Comment by lbdroid** [comment]
If I did, it wasn't intentional.
My socket permissions part is here:
https://github.com/lbdroid/AOSP-SU-PATCH/blob/master/su.patch#L857 through 859

**Comment by phhusson** [comment]
Ok thanks, I guess it's my implementation of mlstrustedobject which doesn't work then.

**Comment by lbdroid** [comment]
https://github.com/lbdroid/AOSP-SU-PATCH/blob/master/su.patch#L328

**Comment by phhusson** [comment]
Yup I've seen it too.
I'm setting all my domains as mlstrustedobject to be sure ATM ;)

**Comment by lbdroid** [comment]
Make sure you keep in mind the difference between mlstrustedobject and mlstrustedsubject...

**Comment by lbdroid** [comment]
The scary thing is that you are asking me questions about selinux.. 5 days ago, selinux was a complete enigma to me.

**Comment by lbdroid** [comment]
FWIW: I just built HEAD and everything is still working for me. The only change I made was packageName since I'm working out of eclipse, so its not taking properly.

**Comment by phhusson** [comment]
Thanks for the info. I'm definitely not setting mlstrustedobject properly...

**Comment by phhusson** [comment]
Ok, I'm pretty sure I've fixed mlstrustedobject, but I still hit the same problem, here is the seinfo:
http://paste.debian.net/318105/
Please note I've changed domain names back to your naming convention
We see su/su_exec/su_device/untrusted_app are mlstrustedobject/mlstrustedsubject (yes I've been as wide as can be)
The error is:
10-25 23:53:31.429  3325  3325 W su      : type=1400 audit(0.0:84): avc: denied { connectto } for path=005355504552555345520000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:r:su:s0 tclass=unix_stream_socket permissive=0
suinfo tells me the rule is right here:
allow untrusted_app su:unix_stream_socket { connectto getopt };

Can you take a quick look to see if you find any obvious error?

**Comment by lbdroid** [comment]
Hmm... how did your socket become su domain? Shouldn't it be su_device?

**Comment by phhusson** [comment]
Well that was part of my question about abstract socket. Current code is
using an abstract socket, not /dev/packagname/xxx, so its context is the
context of the process, not od the (non-existing) file.

**Comment by lbdroid** [comment]
Hmm, so we are running different code then... where is the code pertaining to abstract socket?

**Comment by lbdroid** [comment]
I wonder if that is the same problem as for transitions?

**Comment by phhusson** [comment]
grep SUPERUSER jni/su/su.c
Two memcpy lines (one for client one for sserver)
You can switch from one to the other by commenting the memcpy and the
memset before it.

**Comment by lbdroid** [comment]
Are you sure you actually committed that? I can't find it.
https://github.com/phhusson/Superuser/blob/master/Superuser/jni/su/su.c

**Comment by lbdroid** [comment]
If only we had a newer kernel..
http://www.spinics.net/lists/selinux/msg15312.html

**Comment by lbdroid** [comment]
This is what we *actually* get:

https://android.googlesource.com/kernel/msm/+/android-msm-shamu-3.10-marshmallow/security/selinux/hooks.c#2136

which actually does match upstream...
https://git.kernel.org/cgit/linux/kernel/git/stable/linux-stable.git/tree/security/selinux/hooks.c?id=refs/tags/v3.10.91#n2061

**Comment by phhusson** [comment]
Sorry, it's actually in daemon.c, not su.c
Nice patch, but indeed too bad we can't have it (or we do kernel patching... perhaps later...)

**Comment by lbdroid** [comment]
Wow, that begs the question of why the heck that code isn't activating on my builds? I *definitely* have that code in my source.

**Comment by phhusson** [comment]
F*CK.
Testing again on a 5.1 CAF... rootfs can't store selinux contexts...
SELinux: Could not set context for /sbin/su:  Operation not supported on transport endpoint
I'll try to see if there is an easy option of mount to enable selabels...

**Comment by lbdroid** [comment]
Well... at least <=5.1 should still work the old way. Without dm-verity, the necessity of maintaining an unmodified system image is a lot lower. Unfortunately, this adds a requirement for a separate enablement path that involves a transition (init, su_exec, su). It should be fairly trivial to add that to a boot script.

**Comment by phhusson** [comment]
This makes two totally different install methods.
I think that at least for the moment (ie until I am fully enforced on M) I'll just run with init/init_shell/untrusted_app permissive.

**Comment by lbdroid** [comment]
It isn't actually that different though;

6:
extract boot image,
fix sepolicy,
copy su binary to /sbin/
append service daemonsu to /sbin/init.rc
rebuild boot image,
install boot image,
reboot

5:
extract boot image,
fix sepolicy,
** mount /system
copy su binary to /system/xbin/
setcon the su binary
** umount /system
append service daemonsu to /sbin/init.rc (different path)
rebuild boot image,
install boot image,
reboot

**Comment by lbdroid** [comment]
And we can identify the android version from the ro.bootimage.build.fingerprint line of the /default.prop, or the similar fingerprint line of the /system/build.prop

**Comment by phhusson** [comment]
That would require disabling dm-verity, eg on galaxy s6.
Also I'd like to be able to work only with boot.img (remember I still want :
- easier to distribute (ok init.rc can copy the su binary to system partition itself)
- easier to have a server which automatically generates updated rooting boot.img for many devices

**Comment by lbdroid** [comment]
How about putting the su binary on the userdata partition?

**Comment by phhusson** [comment]
Sounds good to me

**Comment by lbdroid** [comment]
We can protect the integrity (as well as automate update) of the su binary on userdata by comparing its md5sum with the one on the rootfs, and overwriting it if it differs. Do that just prior to actually executing the daemon, and even if it gets changed later on, the privileged one in memory will remain intact.

**Comment by osnapitsjoey** [comment]
Just wanted to let you guys know you're awesome. Can't wait to donate some cash once this is out.

**Comment by phhusson** [comment]
I think that using filename-based transitions we should be able to fix this Android 5.1 problem.

Before thinking of this, I made a big comment abuot what we could do, that might be interesting for other reasons, though it's now false.
I keep it for reference:

It would be actually be an interesting design,
having the ""su"" in in boot.img ONLY do:
- Have itself a su_starter context
- Check a public key on /data/data/me.phh.superuser/files/su
- Change its file context to su_exec
- Run it (and we have a transition su_starter => su)

But it's mostly useless because I believe most of the interesting stuff will be in the sepolicy changes, not in SU itself.
And the problem will be, how apps call our su? We could set PATH to /data/data/... assuming PATH stays alive all the way down to the application, but that would probably kill some apps. (ok we already kill some by being in /sbin/su instead of /system/xbin/su)

We already changed changed /sbin from 0750 to 0755. Would adding an executable right on rootfs from untrusted_app really hurt? (Because that's the only real reason to need a label on su exec)

**Comment by phhusson** [comment]
filename-based transitions are ONLY for when a context creates a new file, it will automatically set the new context, nothing more.
I'm not seeing any neverallow for rootfs execute, and it already has full read access, so I don't see any there could be any security problem with it being also able to execute.

So I'm changing the policy to have executable right on rootfs, su --daemon service gets a selabel, and everyone is happy.

**Comment by lbdroid** [comment]
I was just about to mention that.

**Comment by lbdroid** [comment]
I'm confused about your issue with executing... is it actually not executing the file? Or is a matter of context transitions since you can't set the su_exec context on /sbin/su?

**Comment by lbdroid** [comment]
Don't forget that you can set your transition as a parameter in the init.rc like this;
https://github.com/lbdroid/AOSP-SU-PATCH/blob/master/su.patch#L937

**Comment by lbdroid** [comment]
Oh nevermind. I think I understand what's going on. The 5.1 sepolicy is blocking untrusted_app from executing on rootfs altogether.

**Comment by phhusson** [comment]
If you had to set su_exec mlstrustedobject, it means I'll have to set rootfs mlstrustedobject... :/

**Comment by lbdroid** [comment]
What possible harm could be done to or with rootfs by untrusted_app if it was? After all it is read-only, and even the ram version of it is mounted read-only.

**Comment by lbdroid** [comment]
Hey FYI: I have someone running now with verity turned *ON*. No boot warnings.
We are going to see about convincing android pay to work, which is apparently detecting the su binary. Hopefully it is as simple as renaming the binary to something else.

**Comment by phhusson** [comment]
I have contacted the guy from Android Pay who posted on xda about it, they
started internal discussion about this possibility.

When you say verity, you mean with self-owned keystore+signed boot.img? No
need for dm-verity I guess? (i can't see how the bootloader would know
about dm-verity)
No warning at all in this state worries me a lot...
How to be sure the keystore hasn't changed?

**Comment by lbdroid** [comment]
I mean that he is using the factory system image on a nexus 6, but with root modified into the boot image.

According to a Google document you posted in another issue, nexus 6 and 9 *do not qualify* for verified boot on Android 6, since they both shipped without verified boot on a previous version, making it impossible to prove the boot image valid. The message you got on your nexus 9 when you overrode and enabled verified boot is consistent with it not being a supported operation for end user.

Now 5x and 6p on the other hand have the whole works enabled at factory, so this wouldn't work. Only option there for best state is user signed.

**Comment by lbdroid** [comment]
To continue from #15, that isn't what I meant at all. Of course the user should pick their own password. What I meant is that the su_sensitive domain probably should only be available as a transition from shell user and context, so that some dimwitted application doesn't have any means of getting into it.

**Comment by phhusson** [comment]
Ok understood. I don't think I agree, but we have to fill Policy.md file including with example apps for every cases.
For instance, having su_sensitive only from shell means no FlashFire

**Comment by lbdroid** [comment]
Hmm. Maybe something like this for application access... Password request at the start of *every* session, then step through every command entered with an allow/deny button.

**Comment by phhusson** [comment]
libsuperuser, the standard client implementantion of su is using interactive shells, so doing so would be complicated, and most likely insecure.
Also, most apps usually just su - /my/script
And, though I agree it could be a nice setting for power users, basic root users just won't understand the commands anyway, and will just make them click "agree" faster.

**Comment by lbdroid** [comment]
If the user can't validate the commands, then there are two possible options; don't give them the option at all (shell only), or do, in which case at least we gave them the option to hit the deny button when they didn't understand the command.

I don't actually think it would be that hard to implement. Scan the stream for \n, pop up allow/deny before letting it get through.

**Comment by phhusson** [comment]
\o/ I finally fixed this damn mlstrusted{object,subject} issue in sepolicy-inject in https://github.com/phhusson/sepolicy-inject/commit/370951ab678bb7ab247bc5b607ef5007d09649bb

I am back on tracks for doing a proper binary policy.

**Comment by lbdroid** [comment]
Nice!

**Comment by phhusson** [comment]
Now I have a full-enforcing su \o/
It can do pretty much nothing (launch a shell, dmesg -c), but well it's a start.
My next steps are testing CTS, fixing some bugs this policy made me find.

I'm not sure what to do after that though. Trying to build a script to parse your patch and generate matching rules? Or whitelist by hand(a changed audit2allow might help) while testing several apps?

**Comment by lbdroid** [comment]
I would try by hand. Or rather, manually translate it into your script. I think it really boils down to about 2 dozen lines.

**Comment by lbdroid** [comment]
This should be adequate based on my patch (I haven't yet incorporated your su_daemon domain);
```
New type define:
type su_device, dev_type, mlstrustedobject;

Alter type define:
-type su_exec, exec_type, file_type;
+type su_exec, exec_type, file_type, mlstrustedobject;

New transitions:
domain_auto_trans(init, su_exec, su)
file_type_auto_trans(su, device, su_device)

New allows:
allow init su_exec:file relabelto;
allow domain untrusted_app_devpts:chr_file { open getattr read write ioctl };
allow su self:capability_class_set *;
allow su kernel:system *;
allow su self:memprotect *;
allow su { domain -kernel }:process *;
allow su domain:fd *;
allow su domain:dir *;
allow su domain:lnk_file *;
allow su domain:{ fifo_file file } *;
allow su domain:socket_class_set *;
allow su domain:ipc_class_set *;
allow su domain:key *;
allow su fs_type:filesystem *;
allow su {fs_type dev_type file_type}:dir_file_class_set *;
allow su node_type:node *;
allow su node_type:{ tcp_socket udp_socket rawip_socket } *;
allow su netif_type:netif *;
allow su port_type:socket_class_set *;
allow su port_type:{ tcp_socket dccp_socket } *;
allow su domain:peer *;
allow su domain:binder *;
allow su property_type:property_service *;
allow su service_manager_type:service_manager *;
allow su keystore:keystore_key *;
allow su domain:debuggerd *;
allow su domain:drmservice *;
allow su unlabeled:filesystem *;
allow untrusted_app su_device:sock_file rw_file_perms;
allow untrusted_app su_device:dir r_dir_perms;
allow untrusted_app su_exec:file execute_no_trans;

From AOSP userdebug:
allow adbd self:process setcurrent;
allow adbd su:process dyntransition;
allow adbd adb_data_file:dir rw_dir_perms;
allow adbd adb_data_file:file create_file_perms;
allow domain su:unix_stream_socket connectto;
allow domain su:fd use;
allow domain su:unix_stream_socket { getattr getopt read write shutdown };
binder_call({ domain -init }, su)
allow domain su:fifo_file { write getattr };
allow domain su:process sigchld;
unix_socket_send(wpa, wpa, su)
```
